src/memberAPI.ts

```diff
@@ -1,70 +1,220 @@
-import { checkLoggedIn } from "./api";
-import { db } from "./firebase";
-import { PermissionsManager } from "./permissions";
+import { checkLoggedIn } from './api';
+import { db } from './firebase';
+import { PermissionsManager } from './permissions';
+import { Request, Response } from 'express';
+import { ErrorResponse, MemberResponse, AllMembersResponse } from './APITypes';
+import { Member } from './DataTypes';
+import { readJsonConfigFile } from 'typescript';
 
-export type Member = {
-  email: string,
-  first_name: string,
-  last_name: string,
-  role: string
-}
-
-export let allMembers = async (req, res) => {
+export let allMembers = async (req, res): Promise<AllMembersResponse> => {
   if (checkLoggedIn(req, res)) {
-    res.json({
-      members: await db.collection('members').get().then((vals) => {
-        return vals.docs.map(doc => {
-          return doc.data();
-        });
-      })
-    });
+    let members: Member[] = await db
+      .collection('members')
+      .get()
+      .then((vals) => {
+        return vals.docs.map((doc) => doc.data()) as Member[];
+      });
+    return { members: members, status: 200 };
   }
-}
+};
 
-export let setMember = async (req, res) => {
+export let setMember = async (
+  req,
+  res
+): Promise<MemberResponse | ErrorResponse> => {
   if (checkLoggedIn(req, res)) {
-    let member = await (await db.doc('members/' + req.session.email).get()).data();
+    let member = await (
+      await db.doc('members/' + req.session.email).get()
+    ).data();
     if (!member) {
-      res.status(200).json({ error: "No member with email: " + req.session.email });
+      return {
+        error: 'No member with email: ' + req.session.email,
+        status: 401,
+      };
     } else {
       let canEdit = PermissionsManager.canEditMembers(member.role);
       if (!canEdit) {
-        res.status(200).json({ error: "User with email: " + req.session.email + " does not have permission to edit members!" });
+        return {
+          error:
+            'User with email: ' +
+            req.session.email +
+            ' does not have permission to edit members!',
+          status: 403,
+        };
+      } else {
+        if (!req.body.email || req.body.email === '') {
+          return {
+            error: "Couldn't edit user with undefined email!",
+            status: 400,
+          };
+        }
+        db.doc('members/' + req.body.email)
+          .set(req.body)
+          .then(() => {
+            return { status: 200, member: req.body };
+          })
+          .catch((reason) => {
+            return {
+              error: "Couldn't edit user for reason: " + reason,
+              status: 500,
+            };
+          });
```

Comment:
Like you mentioned earlier, this would have to be updated to correctly return the response as
```
let response = db.doc(...
return response;
```